timelock-impl/src/main/java/com/palantir/atlasdb/backup/AtlasBackupResource.java

```diff
@@ -0,0 +1,178 @@
+    private final Function<String, AsyncTimelockService> timelockServices;
+    private final ConjureResourceExceptionHandler exceptionHandler;
+
+    @VisibleForTesting
+    AtlasBackupResource(
+            RedirectRetryTargeter redirectRetryTargeter, Function<String, AsyncTimelockService> timelockServices) {
+        this.exceptionHandler = new ConjureResourceExceptionHandler(redirectRetryTargeter);
+        this.timelockServices = timelockServices;
+    }
+
+    public static UndertowService undertow(
+            RedirectRetryTargeter redirectRetryTargeter, Function<String, AsyncTimelockService> timelockServices) {
+        return AtlasBackupServiceEndpoints.of(new AtlasBackupResource(redirectRetryTargeter, timelockServices));
+    }
+
+    public static AtlasBackupService jersey(
+            RedirectRetryTargeter redirectRetryTargeter, Function<String, AsyncTimelockService> timelockServices) {
+        return new JerseyAtlasBackupServiceAdapter(new AtlasBackupResource(redirectRetryTargeter, timelockServices));
+    }
+
+    @Override
+    public ListenableFuture<PrepareBackupResponse> prepareBackup(AuthHeader authHeader, PrepareBackupRequest request) {
+        return handleExceptions(() -> Futures.immediateFuture(prepareBackupInternal(request)));
+    }
+
+    private PrepareBackupResponse prepareBackupInternal(PrepareBackupRequest request) {
+        Set<BackupToken> preparedBackups = request.getNamespaces().stream()
+                .map(this::prepareBackup)
+                .flatMap(Optional::stream)
+                .collect(Collectors.toSet());
+        return PrepareBackupResponse.of(preparedBackups);
+    }
+
+    Optional<BackupToken> prepareBackup(Namespace namespace) {
+        try {
+            return Optional.of(tryPrepareBackup(namespace));
+        } catch (Exception ex) {
+            log.info("Failed to prepare backup for namespace", SafeArg.of("namespace", namespace), ex);
+            return Optional.empty();
+        }
+    }
+
+    private BackupToken tryPrepareBackup(Namespace namespace) {
+        AsyncTimelockService timelock = timelock(namespace);
+        LockImmutableTimestampResponse response = timelock.lockImmutableTimestamp(IdentifiedTimeLockRequest.create());
+        long timestamp = timelock.getFreshTimestamp();
+        return BackupToken.builder()
+                .namespace(namespace)
+                .lockToken(response.getLock())
+                .immutableTimestamp(response.getImmutableTimestamp())
+                .backupStartTimestamp(timestamp)
+                .build();
+    }
+
+    @Override
+    public ListenableFuture<CompleteBackupResponse> completeBackup(
+            AuthHeader authHeader, CompleteBackupRequest request) {
+        Map<BackupToken, ListenableFuture<Optional<BackupToken>>> futureMap =
+                request.getBackupTokens().stream().collect(Collectors.toMap(token -> token, this::completeBackupAsync));
```

Comment:
```suggestion
        Map<BackupToken, ListenableFuture<Optional<BackupToken>>> futureMap = KeyedStream.of(request.getBackupTokens())
                .map(this::completeBackupAsync)
                .collectToMap();
```

Reply:
I don't think that works (it certainly doesn't compile) - after the `.map` operation, you just have a `Stream<ListenableFuture<Optional<BackupToken>>>` and no reference to the original token.